common/src/main/java/com/wynntils/gui/render/MapRenderer.java

```diff
@@ -99,12 +109,137 @@ public static void renderMapQuad(
         bufferBuilder.end();
         BufferUploader.end(bufferBuilder);
 
+        float mapTopX = centerX - halfRenderedWidth;
+        float mapLeftZ = centerZ - halfRenderedHeight;
+        float mapBottomX = centerX + halfRenderedWidth;
+        float mapRightZ = centerZ + halfRenderedHeight;
+
+        List<ServicePoi> servicePois = MapModel.getAllPois().stream()
+                .filter(poi -> {
+                    double distanceX = McUtils.player().getX() - mapCenterX;
+                    double distanceZ = McUtils.player().getZ() - mapCenterZ;
+
+                    float textureXPosition = (float) (centerX + distanceX / scale);
+                    float textureZPosition = (float) (centerZ + distanceZ / scale);
+
+                    boolean inMap = textureXPosition >= mapTopX
+                            && textureXPosition <= mapBottomX
+                            && textureZPosition >= mapLeftZ
+                            && textureZPosition <= mapRightZ;
+
+                    return poi instanceof ServicePoi && inMap;
```

Comment:
Recommend that you a) start by testing the instanceof to quickly abort if not relevant, and b) extracting the inMap test in a method.